mutiny-core/src/lsp/lsps.rs

```diff
@@ -0,0 +1,391 @@
+
+use crate::{
+    error::MutinyError,
+    keymanager::PhantomKeysManager,
+    ldkstorage::PhantomChannelManager,
+    logging::MutinyLogger,
+    lsp::{FeeRequest, InvoiceRequest, Lsp, LspConfig},
+    node::{parse_peer_info, LiquidityManager},
+    storage::MutinyStorage,
+    utils,
+};
+
+use super::FeeResponse;
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Clone)]
+pub struct LspsConfig {
+    pub connection_string: String,
+    pub token: Option<String>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct JitChannelInfo {
+    pub channel_id: u128,
+    pub fee_params: OpeningFeeParams,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct GetInfoResponse {
+    pub jit_channel_id: u128,
+    pub opening_fee_params_menu: Vec<OpeningFeeParams>,
+}
+
+pub(crate) struct PendingPaymentInfo {
+    pub expected_fee_msat: Option<u64>,
+    pub fee_params: OpeningFeeParams,
+}
+
+#[derive(Clone)]
+pub struct LspsClient<S: MutinyStorage> {
+    pub pubkey: PublicKey,
+    pub connection_string: String,
+    pub token: Option<String>,
+    liquidity_manager: Arc<LiquidityManager<S>>,
+    channel_manager: Arc<PhantomChannelManager<S>>,
+    keys_manager: Arc<PhantomKeysManager<S>>,
+    network: Network,
+    logger: Arc<MutinyLogger>,
+    pending_fee_requests: Arc<Mutex<HashMap<u128, oneshot::Sender<GetInfoResponse>>>>,
+    pending_buy_requests: Arc<Mutex<HashMap<u128, oneshot::Sender<Bolt11Invoice>>>>,
+    pending_channel_info: Arc<Mutex<HashMap<u128, JitChannelInfo>>>,
+    pending_payments: Arc<Mutex<HashMap<PaymentHash, PendingPaymentInfo>>>,
+}
+
+impl<S: MutinyStorage> LspsClient<S> {
+    pub(crate) fn new(
+        lsp_connection_string: String,
+        token: Option<String>,
+        liquidity_manager: Arc<LiquidityManager<S>>,
+        channel_manager: Arc<PhantomChannelManager<S>>,
+        keys_manager: Arc<PhantomKeysManager<S>>,
+        network: Network,
+        logger: Arc<MutinyLogger>,
+    ) -> Result<Self, MutinyError> {
+        let (lsp_pubkey, _) = parse_peer_info(&lsp_connection_string)?;
+
+        let client = LspsClient {
+            pubkey: lsp_pubkey,
+            connection_string: lsp_connection_string,
+            token,
+            liquidity_manager,
+            channel_manager,
+            keys_manager,
+            network,
+            logger,
+            pending_fee_requests: Arc::new(Mutex::new(HashMap::new())),
+            pending_buy_requests: Arc::new(Mutex::new(HashMap::new())),
+            pending_channel_info: Arc::new(Mutex::new(HashMap::new())),
+            pending_payments: Arc::new(Mutex::new(HashMap::new())),
+        };
+
+        let events_client = client.clone();
+        utils::spawn(async move {
+            events_client.handle_events().await;
+        });
+
+        Ok(client)
+    }
+
+    pub(crate) async fn handle_events(&self) {
+        loop {
+            for event in self.liquidity_manager.get_and_clear_pending_events() {
+                match event {
```

Comment:
Would be nice to move this event handling into a separate function that returns `Result` so we don't need all these `unwraps` and can properly handle the errors

Reply:
did this but doesn't need to return Result itself because really it's sending the Result over the channel to be handled. there's nothing to be done here other than log.